File: mobile/lib/features/colorimetry/engine/color_math.dart
/// Colour-space mathematics for the on-device colorimetric reader.
///
/// This is a deliberate, line-for-line mirror of
/// `backend/app/services/colorimetry/color_math.py`. The two implementations
/// are checked against the same golden vectors in
/// `docs/colorimetry/golden_vectors.json`, so they cannot silently diverge —
/// and if they ever do, a test fails rather than a reading quietly changing
/// depending on whether it was computed on the handset or on the server.
///
/// Everything here is pure and deterministic: no I/O, no platform channels, no
/// dependence on device state.
library;

import 'dart:math' as math;

/// sRGB (IEC 61966-2-1) to CIE XYZ, D65.
const List<List<double>> _srgbToXyz = <List<double>>[
  <double>[0.4124564, 0.3575761, 0.1804375],
  <double>[0.2126729, 0.7151522, 0.0721750],
  <double>[0.0193339, 0.1191920, 0.9503041],
];

/// D65 white point, derived from the matrix above rather than written out as
/// the nominal (95.047, 100.000, 108.883).
///
/// The published matrix coefficients are rounded and do not sum exactly to that
/// triple. Using the nominal value leaves white at L* = 100.0000039 and gives
/// neutral greys a chroma of ~4e-6 — small, but it means "is this patch
/// neutral?" is answered against an inconsistent reference. Deriving the white
/// point from the same matrix makes white exact by construction.
final List<double> d65White = <double>[
  (_srgbToXyz[0][0] + _srgbToXyz[0][1] + _srgbToXyz[0][2]) * 100.0,
  (_srgbToXyz[1][0] + _srgbToXyz[1][1] + _srgbToXyz[1][2]) * 100.0,
  (_srgbToXyz[2][0] + _srgbToXyz[2][1] + _srgbToXyz[2][2]) * 100.0,
];

const double _cieEpsilon = 216.0 / 24389.0;
const double _cieKappa = 24389.0 / 27.0;

/// A CIE L*a*b* colour.
class Lab {
  const Lab(this.l, this.a, this.b);

  final double l;
  final double a;
  final double b;

  double get chroma => math.sqrt(a * a + b * b);

  double get hueDegrees {
    final double h = _degrees(math.atan2(b, a));
    return h < 0 ? h + 360.0 : h;
  }

  List<double> toList() => <double>[l, a, b];

  Lab operator +(Lab other) => Lab(l + other.l, a + other.a, b + other.b);

  Lab operator -(Lab other) => Lab(l - other.l, a - other.a, b - other.b);

  Lab operator *(double k) => Lab(l * k, a * k, b * k);

  /// Euclidean distance in Lab. Used for geometric projection onto the
  /// calibration path, never for reporting a colour difference.
  double distanceTo(Lab other) {
    final double dl = l - other.l;
    final double da = a - other.a;
    final double db = b - other.b;
    return math.sqrt(dl * dl + da * da + db * db);
  }

  @override
  String toString() =>
      'Lab(${l.toStringAsFixed(3)}, ${a.toStringAsFixed(3)}, ${b.toStringAsFixed(3)})';
}

double _degrees(double radians) => radians * 180.0 / math.pi;

double _radians(double degrees) => degrees * math.pi / 180.0;

/// Undo the sRGB transfer function. Input and output in [0, 1].
double srgbToLinear(double channel) {
  return channel <= 0.04045
      ? channel / 12.92
      : math.pow((channel + 0.055) / 1.055, 2.4).toDouble();
}

/// Apply the sRGB transfer function. Input and output in [0, 1].
double linearToSrgb(double channel) {
  final double c = channel.clamp(0.0, double.infinity).toDouble();
  return c <= 0.0031308 ? c * 12.92 : 1.055 * math.pow(c, 1 / 2.4).toDouble() - 0.055;
}

/// sRGB in [0, 1] to CIE XYZ with Y in [0, 100].
List<double> srgbToXyz(double r, double g, double b) {
  final double lr = srgbToLinear(r);
  final double lg = srgbToLinear(g);
  final double lb = srgbToLinear(b);
  return <double>[
    (_srgbToXyz[0][0] * lr + _srgbToXyz[0][1] * lg + _srgbToXyz[0][2] * lb) * 100.0,
    (_srgbToXyz[1][0] * lr + _srgbToXyz[1][1] * lg + _srgbToXyz[1][2] * lb) * 100.0,
    (_srgbToXyz[2][0] * lr + _srgbToXyz[2][1] * lg + _srgbToXyz[2][2] * lb) * 100.0,
  ];
}

double _f(double ratio) => ratio > _cieEpsilon
    ? math.pow(ratio.clamp(0.0, double.infinity), 1.0 / 3.0).toDouble()
    : (_cieKappa * ratio + 16.0) / 116.0;

/// CIE XYZ to CIE L*a*b*.
Lab xyzToLab(List<double> xyz) {
  final double fx = _f(xyz[0] / d65White[0]);
  final double fy = _f(xyz[1] / d65White[1]);
  final double fz = _f(xyz[2] / d65White[2]);
  return Lab(116.0 * fy - 16.0, 500.0 * (fx - fy), 200.0 * (fy - fz));
}

/// CIE L*a*b* back to CIE XYZ.
List<double> labToXyz(Lab lab) {
  final double fy = (lab.l + 16.0) / 116.0;
  final double fx = fy + lab.a / 500.0;
  final double fz = fy - lab.b / 200.0;

  double inv(double t) {
    final double t3 = t * t * t;
    return t3 > _cieEpsilon ? t3 : (116.0 * t - 16.0) / _cieKappa;
  }

  return <double>[
    inv(fx) * d65White[0],
    inv(fy) * d65White[1],
    inv(fz) * d65White[2],
  ];
}

/// 8-bit sRGB straight to L*a*b*.
Lab srgbU8ToLab(double r, double g, double b) =>
    xyzToLab(srgbToXyz(r / 255.0, g / 255.0, b / 255.0));

/// L*a*b* to sRGB in [0, 1], clipped to the displayable gamut.
List<double> labToSrgb(Lab lab) {
  final List<double> xyz = labToXyz(lab);
  final double x = xyz[0] / 100.0;
  final double y = xyz[1] / 100.0;
  final double z = xyz[2] / 100.0;

  // Inverse of _srgbToXyz, computed once as constants so the Dart and Python
  // implementations use identical numbers.
  const double m00 = 3.2404542, m01 = -1.5371385, m02 = -0.4985314;
  const double m10 = -0.9692660, m11 = 1.8760108, m12 = 0.0415560;
  const double m20 = 0.0556434, m21 = -0.2040259, m22 = 1.0572252;

  final double lr = (m00 * x + m01 * y + m02 * z).clamp(0.0, 1.0).toDouble();
  final double lg = (m10 * x + m11 * y + m12 * z).clamp(0.0, 1.0).toDouble();
  final double lb = (m20 * x + m21 * y + m22 * z).clamp(0.0, 1.0).toDouble();

  return <double>[
    linearToSrgb(lr).clamp(0.0, 1.0).toDouble(),
    linearToSrgb(lg).clamp(0.0, 1.0).toDouble(),
    linearToSrgb(lb).clamp(0.0, 1.0).toDouble(),
  ];
}

/// CIE76 Euclidean distance. Coarse gating only, never reported.
double deltaE76(Lab a, Lab b) => a.distanceTo(b);

/// CIEDE2000 colour difference.
///
/// Follows Sharma, Wu & Dalal (2005), including the hue-discontinuity handling
/// that naive implementations get wrong. The Python reference is checked
/// against that paper's 34 published test pairs; this port is checked against
/// the same values through the shared golden vectors.
double deltaE2000(
  Lab lab1,
  Lab lab2, {
  double kL = 1.0,
  double kC = 1.0,
  double kH = 1.0,
}) {
  final double c1 = math.sqrt(lab1.a * lab1.a + lab1.b * lab1.b);
  final double c2 = math.sqrt(lab2.a * lab2.a + lab2.b * lab2.b);
  final double cBar = (c1 + c2) / 2.0;

  final double cBar7 = math.pow(cBar, 7).toDouble();
  final double pow25_7 = math.pow(25.0, 7).toDouble();
  final double g = 0.5 * (1.0 - math.sqrt(cBar7 / (cBar7 + pow25_7)));

  final double a1p = (1.0 + g) * lab1.a;
  final double a2p = (1.0 + g) * lab2.a;
  final double c1p = math.sqrt(a1p * a1p + lab1.b * lab1.b);
  final double c2p = math.sqrt(a2p * a2p + lab2.b * lab2.b);

  double hue(double ap, double bp) {
    if (ap.abs() < 1e-12 && bp.abs() < 1e-12) return 0.0;
    final double h = _degrees(math.atan2(bp, ap));
    return h < 0 ? h + 360.0 : h;
  }

  final double h1p = hue(a1p, lab1.b);
  final double h2p = hue(a2p, lab2.b);

  final double dLp = lab2.l - lab1.l;
  final double dCp = c2p - c1p;

  final bool chromaProductZero = (c1p * c2p) == 0;
  final double dhpRaw = h2p - h1p;
  double dhp;
  if (chromaProductZero) {
    dhp = 0.0;
  } else if (dhpRaw.abs() <= 180.0) {
    dhp = dhpRaw;
  } else if (dhpRaw > 180.0) {
    dhp = dhpRaw - 360.0;
  } else {
    dhp = dhpRaw + 360.0;
  }
  final double dHp = 2.0 * math.sqrt(c1p * c2p) * math.sin(_radians(dhp / 2.0));

  final double lpBar = (lab1.l + lab2.l) / 2.0;
  final double cpBar = (c1p + c2p) / 2.0;

  final double hSum = h1p + h2p;
  final double hAbs = (h1p - h2p).abs();
  double hpBar;
  if (chromaProductZero) {
    hpBar = hSum;
  } else if (hAbs <= 180.0) {
    hpBar = hSum / 2.0;
  } else if (hSum < 360.0) {
    hpBar = (hSum + 360.0) / 2.0;
  } else {
    hpBar = (hSum - 360.0) / 2.0;
  }

  final double t = 1.0 -
      0.17 * math.cos(_radians(hpBar - 30.0)) +
      0.24 * math.cos(_radians(2.0 * hpBar)) +
      0.32 * math.cos(_radians(3.0 * hpBar + 6.0)) -
      0.20 * math.cos(_radians(4.0 * hpBar - 63.0));

  final double dTheta = 30.0 * math.exp(-math.pow((hpBar - 275.0) / 25.0, 2).toDouble());
  final double cpBar7 = math.pow(cpBar, 7).toDouble();
  final double rC = 2.0 * math.sqrt(cpBar7 / (cpBar7 + pow25_7));
  final double rT = -rC * math.sin(_radians(2.0 * dTheta));

  final double lpBarSq = (lpBar - 50.0) * (lpBar - 50.0);
  final double sL = 1.0 + (0.015 * lpBarSq) / math.sqrt(20.0 + lpBarSq);
  final double sC = 1.0 + 0.045 * cpBar;
  final double sH = 1.0 + 0.015 * cpBar * t;

  final double termL = dLp / (kL * sL);
  final double termC = dCp / (kC * sC);
  final double termH = dHp / (kH * sH);

  return math.sqrt(
    termL * termL + termC * termC + termH * termH + rT * termC * termH,
  );
}

/// Perceptual mean of a block of 8-bit RGB pixels.
///
/// Averaging happens in linear light, not in gamma-encoded sRGB: the mean of
/// gamma-encoded values is systematically darker than the mean of the light the
/// sensor actually received.
Lab meanLab(List<int> rgbBytes) {
  if (rgbBytes.isEmpty) return const Lab(0, 0, 0);
  final int pixelCount = rgbBytes.length ~/ 3;
  double sumR = 0, sumG = 0, sumB = 0;
  for (int i = 0; i < pixelCount; i++) {
    sumR += srgbToLinear(rgbBytes[i * 3] / 255.0);
    sumG += srgbToLinear(rgbBytes[i * 3 + 1] / 255.0);
    sumB += srgbToLinear(rgbBytes[i * 3 + 2] / 255.0);
  }
  final double lr = sumR / pixelCount;
  final double lg = sumG / pixelCount;
  final double lb = sumB / pixelCount;
  return xyzToLab(<double>[
    (_srgbToXyz[0][0] * lr + _srgbToXyz[0][1] * lg + _srgbToXyz[0][2] * lb) * 100.0,
    (_srgbToXyz[1][0] * lr + _srgbToXyz[1][1] * lg + _srgbToXyz[1][2] * lb) * 100.0,
    (_srgbToXyz[2][0] * lr + _srgbToXyz[2][1] * lg + _srgbToXyz[2][2] * lb) * 100.0,
  ]);
}

File: backend/app/services/colorimetry/color_math.py
"""Colour-space mathematics for the SATVA colorimetric reader.

Everything here is pure, deterministic and free of I/O so that it can be
unit-tested against published reference values and mirrored exactly in the Dart
implementation that runs on the handset. See
`mobile/lib/features/colorimetry/engine/color_math.dart` and the shared golden
vectors in `docs/colorimetry/golden_vectors.json`.

Conventions
-----------
* sRGB values are floats in [0, 1] unless a function name says ``_u8``.
* The reference white is D65 (the illuminant sRGB is defined against).
* ``delta_e_2000`` implements CIEDE2000 (Sharma, Wu & Dalal 2005). The unit
  tests check it against that paper's published test-data table, because a
  subtly wrong ΔE would silently corrupt every reading SATVA reports.
"""

from __future__ import annotations

import math

import numpy as np

# sRGB (IEC 61966-2-1) to CIE XYZ, D65.
_SRGB_TO_XYZ = np.array(
    [
        [0.4124564, 0.3575761, 0.1804375],
        [0.2126729, 0.7151522, 0.0721750],
        [0.0193339, 0.1191920, 0.9503041],
    ],
    dtype=np.float64,
)

# D65 white point, 2-degree standard observer, normalised so Y = 100.
#
# Derived from the matrix above rather than written out as the nominal
# (95.047, 100.000, 108.883), because the published matrix coefficients are
# rounded and do not sum exactly to that triple. Using the nominal value leaves
# white at L* = 100.0000039 and gives neutral greys a chroma of ~4e-6 -- small,
# but it means "is this patch neutral?" is answered against an inconsistent
# reference. Deriving the white point from the same matrix makes white exact by
# construction and neutrals exactly achromatic.
D65_WHITE = _SRGB_TO_XYZ.sum(axis=1) * 100.0

# The nominal published value, kept for documentation and cross-checking.
D65_WHITE_NOMINAL = np.array([95.047, 100.000, 108.883], dtype=np.float64)

_CIE_EPSILON = 216.0 / 24389.0
_CIE_KAPPA = 24389.0 / 27.0


def srgb_to_linear(channel: np.ndarray | float) -> np.ndarray:
    """Undo the sRGB transfer function. Input and output are in [0, 1]."""
    c = np.asarray(channel, dtype=np.float64)
    return np.where(c <= 0.04045, c / 12.92, ((c + 0.055) / 1.055) ** 2.4)


def linear_to_srgb(channel: np.ndarray | float) -> np.ndarray:
    """Apply the sRGB transfer function. Input and output are in [0, 1]."""
    c = np.asarray(channel, dtype=np.float64)
    return np.where(
        c <= 0.0031308, c * 12.92, 1.055 * np.power(np.clip(c, 0, None), 1 / 2.4) - 0.055
    )


def srgb_to_xyz(rgb: np.ndarray) -> np.ndarray:
    """sRGB in [0, 1] -> XYZ with Y in [0, 100]. Accepts (3,) or (N, 3)."""
    arr = np.atleast_2d(np.asarray(rgb, dtype=np.float64))
    linear = srgb_to_linear(arr)
    xyz = linear @ _SRGB_TO_XYZ.T * 100.0
    return xyz[0] if np.ndim(rgb) == 1 else xyz


def xyz_to_lab(xyz: np.ndarray, white: np.ndarray = D65_WHITE) -> np.ndarray:
    """CIE XYZ -> CIE L*a*b*. Accepts (3,) or (N, 3)."""
    arr = np.atleast_2d(np.asarray(xyz, dtype=np.float64))
    ratio = arr / white

    f = np.where(
        ratio > _CIE_EPSILON,
        np.cbrt(np.clip(ratio, 0, None)),
        (_CIE_KAPPA * ratio + 16.0) / 116.0,
    )
    fx, fy, fz = f[:, 0], f[:, 1], f[:, 2]

    lab = np.stack([116.0 * fy - 16.0, 500.0 * (fx - fy), 200.0 * (fy - fz)], axis=1)
    return lab[0] if np.ndim(xyz) == 1 else lab


def lab_to_xyz(lab: np.ndarray, white: np.ndarray = D65_WHITE) -> np.ndarray:
    arr = np.atleast_2d(np.asarray(lab, dtype=np.float64))
    ell, a, b = arr[:, 0], arr[:, 1], arr[:, 2]

    fy = (ell + 16.0) / 116.0
    fx = fy + a / 500.0
    fz = fy - b / 200.0

    def _inv(t: np.ndarray) -> np.ndarray:
        t3 = t**3
        return np.where(t3 > _CIE_EPSILON, t3, (116.0 * t - 16.0) / _CIE_KAPPA)

    xyz = np.stack([_inv(fx), _inv(fy), _inv(fz)], axis=1) * white
    return xyz[0] if np.ndim(lab) == 1 else xyz


def srgb_to_lab(rgb: np.ndarray) -> np.ndarray:
    """Convenience: sRGB in [0, 1] straight to L*a*b*."""
    return xyz_to_lab(srgb_to_xyz(rgb))


def srgb_u8_to_lab(rgb_u8: np.ndarray) -> np.ndarray:
    """Convenience: 8-bit sRGB straight to L*a*b*."""
    return srgb_to_lab(np.asarray(rgb_u8, dtype=np.float64) / 255.0)


def lab_to_srgb(lab: np.ndarray) -> np.ndarray:
    """L*a*b* back to sRGB in [0, 1], clipped to the displayable gamut."""
    xyz = np.atleast_2d(lab_to_xyz(lab)) / 100.0
    linear = xyz @ np.linalg.inv(_SRGB_TO_XYZ).T
    out = np.clip(linear_to_srgb(np.clip(linear, 0.0, 1.0)), 0.0, 1.0)
    return out[0] if np.ndim(lab) == 1 else out


def delta_e_76(lab1: np.ndarray, lab2: np.ndarray) -> float | np.ndarray:
    """CIE76 Euclidean distance. Used only for coarse gating, never reporting."""
    a = np.atleast_2d(np.asarray(lab1, dtype=np.float64))
    b = np.atleast_2d(np.asarray(lab2, dtype=np.float64))
    d = np.sqrt(np.sum((a - b) ** 2, axis=1))
    return float(d[0]) if d.size == 1 else d


def delta_e_2000(
    lab1: np.ndarray,
    lab2: np.ndarray,
    *,
    k_l: float = 1.0,
    k_c: float = 1.0,
    k_h: float = 1.0,
) -> float | np.ndarray:
    """CIEDE2000 colour difference.

    Implementation follows Sharma, Wu & Dalal (2005), "The CIEDE2000
    Color-Difference Formula: Implementation Notes, Supplementary Test Data, and
    Mathematical Observations". The hue-difference discontinuities called out in
    that paper are handled explicitly rather than relying on atan2 alone, which
    is where naive implementations usually diverge.
    """
    a = np.atleast_2d(np.asarray(lab1, dtype=np.float64))
    b = np.atleast_2d(np.asarray(lab2, dtype=np.float64))

    l1, a1, b1 = a[:, 0], a[:, 1], a[:, 2]
    l2, a2, b2 = b[:, 0], b[:, 1], b[:, 2]

    c1 = np.hypot(a1, b1)
    c2 = np.hypot(a2, b2)
    c_bar = (c1 + c2) / 2.0

    c_bar7 = c_bar**7
    g = 0.5 * (1.0 - np.sqrt(c_bar7 / (c_bar7 + 25.0**7)))

    a1p = (1.0 + g) * a1
    a2p = (1.0 + g) * a2
    c1p = np.hypot(a1p, b1)
    c2p = np.hypot(a2p, b2)

    def _hue(ap: np.ndarray, bp: np.ndarray) -> np.ndarray:
        h = np.degrees(np.arctan2(bp, ap))
        h = np.where(h < 0, h + 360.0, h)
        # Convention: hue is 0 when chroma is 0.
        return np.where((np.abs(ap) < 1e-12) & (np.abs(bp) < 1e-12), 0.0, h)

    h1p = _hue(a1p, b1)
    h2p = _hue(a2p, b2)

    dlp = l2 - l1
    dcp = c2p - c1p

    c_prod_zero = (c1p * c2p) == 0
    dhp_raw = h2p - h1p
    dhp = np.where(
        c_prod_zero,
        0.0,
        np.where(
            np.abs(dhp_raw) <= 180.0,
            dhp_raw,
            np.where(dhp_raw > 180.0, dhp_raw - 360.0, dhp_raw + 360.0),
        ),
    )
    dhp_term = 2.0 * np.sqrt(c1p * c2p) * np.sin(np.radians(dhp / 2.0))

    lp_bar = (l1 + l2) / 2.0
    cp_bar = (c1p + c2p) / 2.0

    h_sum = h1p + h2p
    h_abs = np.abs(h1p - h2p)
    hp_bar = np.where(
        c_prod_zero,
        h_sum,
        np.where(
            h_abs <= 180.0,
            h_sum / 2.0,
            np.where(h_sum < 360.0, (h_sum + 360.0) / 2.0, (h_sum - 360.0) / 2.0),
        ),
    )

    t = (
        1.0
        - 0.17 * np.cos(np.radians(hp_bar - 30.0))
        + 0.24 * np.cos(np.radians(2.0 * hp_bar))
        + 0.32 * np.cos(np.radians(3.0 * hp_bar + 6.0))
        - 0.20 * np.cos(np.radians(4.0 * hp_bar - 63.0))
    )

    d_theta = 30.0 * np.exp(-(((hp_bar - 275.0) / 25.0) ** 2))
    cp_bar7 = cp_bar**7
    r_c = 2.0 * np.sqrt(cp_bar7 / (cp_bar7 + 25.0**7))
    r_t = -r_c * np.sin(np.radians(2.0 * d_theta))

    lp_bar_sq = (lp_bar - 50.0) ** 2
    s_l = 1.0 + (0.015 * lp_bar_sq) / np.sqrt(20.0 + lp_bar_sq)
    s_c = 1.0 + 0.045 * cp_bar
    s_h = 1.0 + 0.015 * cp_bar * t

    term_l = dlp / (k_l * s_l)
    term_c = dcp / (k_c * s_c)
    term_h = dhp_term / (k_h * s_h)

    de = np.sqrt(term_l**2 + term_c**2 + term_h**2 + r_t * term_c * term_h)
    return float(de[0]) if de.size == 1 else de


def mean_lab(pixels_u8: np.ndarray) -> np.ndarray:
    """Perceptual mean of an 8-bit RGB pixel block.

    Averaging must happen in a linear space, not in gamma-encoded sRGB: the mean
    of gamma-encoded values is systematically darker than the mean of the light
    the sensor actually received. We therefore average linear RGB and convert the
    result once.
    """
    arr = np.asarray(pixels_u8, dtype=np.float64).reshape(-1, 3) / 255.0
    linear_mean = srgb_to_linear(arr).mean(axis=0)
    xyz = (linear_mean @ _SRGB_TO_XYZ.T) * 100.0
    return xyz_to_lab(xyz)


def chroma(lab: np.ndarray) -> float:
    lab = np.asarray(lab, dtype=np.float64)
    return float(math.hypot(lab[1], lab[2]))


def hue_angle_deg(lab: np.ndarray) -> float:
    lab = np.asarray(lab, dtype=np.float64)
    h = math.degrees(math.atan2(lab[2], lab[1]))
    return h + 360.0 if h < 0 else h
